text/0009-BundleTracker-integration.md

```diff
@@ -0,0 +1,304 @@
+
+PP2: Listeners on their own are insufficient in creating an up-to-date record of bundle states for user code, since bundle states that have not changed since the addition of the listener will not be tracked.
+
+PP3: Leaving the proper tracker implementation to the user creates many opportunities for race conditions or deadlock.
+
+PP4: For a tracker to be useful, it is helpful to include methods of customization and extension, so that custom wrapper objects and callbacks can be included.
+
+### Requirements
+
+<table>
+<thead>
+<tr class="header">
+<th><p>ID</p></th>
+<th><p>Statement</p></th>
+<th><p>Pain Point ID / Rationale</p></th>
+<th><p>Priority</p></th>
+</tr>
+</thead>
+<tbody>
+<tr class="odd">
+<td>R1_Previous_States</td>
+<td>The solution must be able to track bundles started before initialization.</td>
+<td>PP1, PP2</td>
+<td>Must have</td>
+</tr>
+<tr class="even">
+<td>R2_Concurrency</td>
+<td>The solution must handle concurrency when grabbing states of bundles existing during tracker initialization.</td>
+<td>PP3</td>
+<td>Must have</td>
+</tr>
+<tr class="odd">
+<td>R3_Customization</td>
+<td>The solution should allow for the user to customize callbacks through a customizer object or subclassing the BundleTracker.</td>
+<td>PP4</td>
+<td>Must have</td>
+</tr>
+<tr class="even">
+<td>R3_Extension</td>
+<td>The BundleTracker should be generic, such that a wrapper class can be defined to pass additional information through the tracking system.</td>
+<td>PP4</td>
+<td>Must have</td>
+</tr>
+<tr class="odd">
+<td>R5_Standardization</td>
+<td>The solution should map to C++ from the OSGi specification for BundleTracker.</td>
+<td><a href="http://docs.osgi.org/specification/osgi.core/7.0.0/util.tracker.html#d0e52020">OSGi Spec</a></td>
+<td>Must have</td>
+</tr>
+<tr class="even">
+<td>R6_Quality</td>
+<td>The solution should be robust and bug-free. This is doubly crucial as Declarative Services would then depend on the BundleTracker.</td>
+<td>PP1</td>
+<td>Must have</td>
+</tr>
+<tr class="odd">
+<td>R7_Efficiency</td>
+<td>The solution should be able to scale to many bundles without slowing the core framework.</td>
+<td><a href="http://docs.osgi.org/specification/osgi.core/7.0.0/util.tracker.html#d0e52020"></a></td>
+<td>Must have</td>
+</tr>
+</tbody>
+</table>
+
+## Functional design
+
+### Functional design summary
+
+The API for the `BundleTracker` closely mirrors that of the OSGi specification, mapped to an implementation in C++. The `ServiceTracker`, which has been implemented in the core framework, is used as a guide to ensure uniformity. However, certain enhancements are provided (e.g. smart pointers) to the `BundleTracker` API to simplify lifetime and ownership of certain objects.
+
+The `BundleTracker` is templated to allow for custom object types to be tracked, containing extra data or functionality.
```

Comment:
I'd like to see the entire BundleTracker class declaration here instead of it being broken up into chunks. The usage can be moved down after the BundleTracker class declaration has been laid out.